File: 300.600/332.ReconstructItinerary.cpp
#include <iostream>
#include <unordered_map>
#include <queue>
#include <vector>
#include <functional>

using namespace std;

#define START "JFK"

// 332. 重新安排行程
class ReconstructItinerary {
public:
    vector<string> findItinerary(vector<vector<string>>& tickets) {
        unordered_map<string, priority_queue<string, vector<string>, greater<string>>> dp;
        for (auto& vec : tickets) {
            dp[vec[0]].emplace(vec[1]);
        }
        vector<string> ret;
        dfs(dp, ret, START);
        reverse(ret.begin(), ret.end());
        return ret;
    }
private:
    void dfs(unordered_map<string, priority_queue<string, vector<string>, greater<string>>>& dp, vector<string>& ret, const string& curr) {
        while (dp.count(curr) && dp[curr].size() > 0) {
            string c = dp[curr].top();
            dp[curr].pop();
            dfs(dp, ret, c);
        }
        ret.push_back(curr);
    }
};

int main(int argc, char const* argv[]) {
    ReconstructItinerary cls;
    auto& ret = cls.findItinerary(vector<vector<string>>{
        {"MUC", "LHR"}, { "JFK","MUC" }, { "SFO","SJC" }, { "LHR","SFO" }
    });
    cout << "[";
    for (auto& it : ret) {
        cout << it.c_str() << " , ";
    }
    cout << "]" << endl;

    auto& ret1 = cls.findItinerary(vector<vector<string>>{
        {"JFK", "SFO"}, { "JFK","ATL" }, { "SFO","ATL" }, { "ATL","JFK" }, { "ATL","SFO" }
    });
    cout << "[";
    for (auto& it : ret1) {
        cout << it.c_str() << " , ";
    }
    cout << "]" << endl;

    return 0;
}

File: go/300.600/332.ReconstructItinerary.go
package main

import (
	"fmt"
	"sort"
)

func findItinerary(tickets [][]string) []string {
	dp := map[string][]string{}
	for _, v := range tickets {
		dp[v[0]] = append(dp[v[0]], v[1])
	}
	for _, v := range dp {
		sort.Strings(v)
	}
	ret := []string{}
	var dfs func(cur string)
	dfs = func(cur string) {
		v, ok := dp[cur]
		for ok && len(v) > 0 {
			c := v[0]
			dp[cur] = v[1:]
			dfs(c)
			v, ok = dp[cur]
		}
		ret = append(ret, cur)
	}
	dfs("JFK")
	//reverse
	for i := 0; i < len(ret)/2; i++ {
		ret[i], ret[len(ret)-i-1] = ret[len(ret)-i-1], ret[i]
	}
	return ret
}

func main() {
	fmt.Println(findItinerary([][]string{{"MUC", "LHR"}, {"JFK", "MUC"}, {"SFO", "SJC"}, {"LHR", "SFO"}}))
	fmt.Println(findItinerary([][]string{{"JFK", "SFO"}, {"JFK", "ATL"}, {"SFO", "ATL"}, {"ATL", "JFK"}, {"ATL", "SFO"}}))
}
